Range APIs do not construct / move trees in tree order (observable by custom elements)

rniwa opened the pull request:
#### fa5a06a66c96bc6864e684d9f9f07b115f39c60e
<pre>
Range APIs do not construct / move trees in tree order (observable by custom elements)
<a href="https://bugs.webkit.org/show_bug.cgi?id=188279">https://bugs.webkit.org/show_bug.cgi?id=188279</a>

Reviewed by Chris Dumez.

In order to match the spec&apos;ed upgrade order, create a nested custom element reaction stack
which is used as a holding tank of upgrade reactions during Range::processContents.

At the end, we traverse the document fragment top-down and re-enqueue each element to the real
custom element reaction stack in the tree order.

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/range-and-constructors-expected.txt:
* Source/WebCore/dom/CustomElementReactionQueue.cpp:
(WebCore::CustomElementQueue::takeElements): Added.
(WebCore::CustomElementReactionQueue::enqueueElementsOnAppropriateElementQueue): Added.
(WebCore::CustomElementReactionStack::takeElements): Added.
* Source/WebCore/dom/CustomElementReactionQueue.h:
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::processContents):

Canonical link: <a href="https://commits.webkit.org/259987@main">https://commits.webkit.org/259987@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fcd1be1195518d65df5900aaba9f16f71ab35af7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106628 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15625 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39420 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115813 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/115433 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110535 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17133 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6858 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98820 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112397 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13013 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/95978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/94891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27623 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8851 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28977 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9403 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/6016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15018 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48523 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10931 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3715 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/CustomElementReactionQueue.cpp:
```diff
@@ -386,6 +386,18 @@ inline void CustomElementQueue::processQueue(JSC::JSGlobalObject* state)
+Vector<GCReachableRef<Element>> CustomElementQueue::takeElements()
+{
+    RELEASE_ASSERT(!m_invoking);
+    return WTFMove(m_elements);
```
I like std::exchange() better in such cases. The reasoning is that it is normally not safe to assume `WTFMove()` leaves m_elements in a good state (though it does here because of the way we wrote the Vector move constructor).

rniwa replied:
Ok, sure.

cdumez commented on Source/WebCore/dom/CustomElementReactionQueue.cpp:
```diff
@@ -415,6 +427,16 @@ void CustomElementReactionStack::processQueue(JSC::JSGlobalObject* state)
+    if (!m_queue)
+        return { };
+    Vector<GCReachableRef<Element>> elements = m_queue->takeElements();
+    delete m_queue;
```
Not new to this patch but I wish we didn't do explicit memory management here. Ideally m_queue would be a unique_ptr<> for e.g.

rniwa replied:
I know but we can't do it because that would generate call to delete in its destructor & regress perf :(

cdumez commented on Source/WebCore/dom/Range.cpp:
```diff
@@ -317,95 +319,111 @@ ExceptionOr<RefPtr<DocumentFragment>> Range::processContents(ActionType action)
+        RangeBoundaryPoint originalEnd(m_end);
+
+        // what is the highest node that partially selects the start / end of the range?
+        RefPtr<Node> partialStart = highestAncestorUnderCommonRoot(&originalStart.container(), commonRoot.get());
```
Would `RefPtr` suffice instead of `RefPtr<Node>`?

rniwa replied:
Sure.

cdumez commented on Source/WebCore/dom/Range.cpp:
```diff
@@ -317,95 +319,111 @@ ExceptionOr<RefPtr<DocumentFragment>> Range::processContents(ActionType action)
+
+        // what is the highest node that partially selects the start / end of the range?
+        RefPtr<Node> partialStart = highestAncestorUnderCommonRoot(&originalStart.container(), commonRoot.get());
+        RefPtr<Node> partialEnd = highestAncestorUnderCommonRoot(&originalEnd.container(), commonRoot.get());
```
ditto.

rniwa replied:
Sure.

cdumez commented on Source/WebCore/dom/Range.cpp:
```diff
@@ -317,95 +319,111 @@ ExceptionOr<RefPtr<DocumentFragment>> Range::processContents(ActionType action)
+        // after any DOM mutation event, at various stages below. See webkit bug 60350.
+
+        RefPtr<Node> leftContents;
+        if (&originalStart.container() != commonRoot && commonRoot->contains(&originalStart.container())) {
```
Should be calling `contains(originalStart.container())` instead of passing a raw pointer. The overload exists and may avoid a null check.

rniwa replied:
Good catch. Will do.

cdumez commented on Source/WebCore/dom/Range.cpp:
```diff
@@ -317,95 +319,111 @@ ExceptionOr<RefPtr<DocumentFragment>> Range::processContents(ActionType action)
-            rightContents = secondResult.releaseReturnValue();
-    }
+        RefPtr<Node> rightContents;
+        if (&endContainer() != commonRoot && commonRoot->contains(&originalEnd.container())) {
```
ditto.

rniwa replied:
Will do

cdumez commented on Source/WebCore/dom/Range.cpp:
```diff
@@ -317,95 +319,111 @@ ExceptionOr<RefPtr<DocumentFragment>> Range::processContents(ActionType action)
+        }
+
+        // delete all children of commonRoot between the start and end container
+        RefPtr<Node> processStart = childOfCommonRootBeforeOffset(&originalStart.container(), originalStart.offset(), commonRoot.get());
```
`RefPtr` without `<Node>`?

rniwa replied:
Done.

cdumez commented on Source/WebCore/dom/Range.cpp:
```diff
@@ -317,95 +319,111 @@ ExceptionOr<RefPtr<DocumentFragment>> Range::processContents(ActionType action)
+        RefPtr<Node> processStart = childOfCommonRootBeforeOffset(&originalStart.container(), originalStart.offset(), commonRoot.get());
+        if (processStart && &originalStart.container() != commonRoot) // processStart contains nodes before m_start.
+            processStart = processStart->nextSibling();
+        RefPtr<Node> processEnd = childOfCommonRootBeforeOffset(&originalEnd.container(), originalEnd.offset(), commonRoot.get());
```
ditto.

rniwa replied:
Done.

cdumez commented on Source/WebCore/dom/Range.cpp:
```diff
@@ -317,95 +319,111 @@ ExceptionOr<RefPtr<DocumentFragment>> Range::processContents(ActionType action)
+
+        // Collapse the range, making sure that the result is not within a node that was partially selected.
+        if (action == Extract || action == Delete) {
+            if (partialStart && commonRoot->contains(partialStart.get())) {
```
`if (commonRoot->contains(partialStart.get())) {`
or
`if (partialStart && commonRoot->contains(*partialStart)) {`

rniwa replied:
Opted for the latter.

cdumez commented on Source/WebCore/dom/Range.cpp:
```diff
@@ -317,95 +319,111 @@ ExceptionOr<RefPtr<DocumentFragment>> Range::processContents(ActionType action)
+                auto result = setStart(*partialStart->parentNode(), partialStart->computeNodeIndex() + 1);
+                if (result.hasException())
+                    return result.releaseException();
+            } else if (partialEnd && commonRoot->contains(partialEnd.get())) {
```
ditto.

rniwa replied:
Fixed.

cdumez commented on Source/WebCore/dom/Range.cpp:
```diff
@@ -317,95 +319,111 @@ ExceptionOr<RefPtr<DocumentFragment>> Range::processContents(ActionType action)
+                if (result.hasException())
+                    return result.releaseException();
+            }
+            collapse(true);
```
Probably not new to this patch but it'd be nice if this took an enum class instead of a boolean. `true` is kinda obscure here.

rniwa replied:
Yeah, maybe fix it as a follow up.

cdumez reviewed (approved):
LGTM with a few nits.

rniwa committed:
Range APIs do not construct / move trees in tree order (observable by custom elements)
https://bugs.webkit.org/show_bug.cgi?id=188279

Reviewed by Chris Dumez.

In order to match the spec'ed upgrade order, create a nested custom element reaction stack
which is used as a holding tank of upgrade reactions during Range::processContents.

At the end, we traverse the document fragment top-down and re-enqueue each element to the real
custom element reaction stack in the tree order.

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/range-and-constructors-expected.txt:
* Source/WebCore/dom/CustomElementReactionQueue.cpp:
(WebCore::CustomElementQueue::takeElements): Added.
(WebCore::CustomElementReactionQueue::enqueueElementsOnAppropriateElementQueue): Added.
(WebCore::CustomElementReactionStack::takeElements): Added.
* Source/WebCore/dom/CustomElementReactionQueue.h:
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::processContents):

Canonical link: https://commits.webkit.org/259987@main
